circuit_tutorials/plonky2/merkle_tree/src/main.rs

```diff
@@ -0,0 +1,253 @@
+
+#[derive(Deserialize, Debug)]
+pub struct JsonProofData {
+    pub proof: String,
+    pub common: String,
+    pub verifier_data: String,
+}
+
+#[tokio::main]
+async fn main() {
+    // Obtain the user's API key from the .env file.
+    dotenv().expect("Failed to read .env file");
+    let api_key: String = std::env::var("SINDRI_API_KEY").unwrap();
+
+    // Create a headers map with the API key.
+    let header = headers_json(&api_key);
+
+    // Uploads the circuit code to Sindri.
+    compile_circuit(header.clone()).await;
+
+    // Uploads an input to the circuit consisting of a vector of 1024 leaves and an index value.
+    // Proof artifacts are saved as a JSON file in the /data/ directory.
+    let input_path: &str = "input_1024.json";
+    prove_circuit(input_path, header).await;
+
+    // Verifies the proof.
+    let proof_path: &str = "./data/prove_out.json";
+    verify_proof(proof_path);
+}
+
+// This function uploads the circuit to Sindri for compilation and outputs the circuit details as 
+// a JSON file in the the /data/ directory.
+async fn compile_circuit(header: HeaderMap) {
+    let mut contents = Vec::new();
+    // This block is scoped that contents can be accessed after written to.
+    {
+        let buffer = std::io::Cursor::new(&mut contents);
+        let enc = GzEncoder::new(buffer, Compression::default());
+        let mut tar = tar::Builder::new(enc);
+        tar.append_dir_all("merkle_tree", "./circuit").unwrap();
+    }
+
+    let part = Part::bytes(contents).file_name("filename.filetype");
+    let upload = reqwest::multipart::Form::new().part("files", part);
+
+    // Create a new circuit.
+    println!("Compiling circuit");
+    let client = Client::new();
+    let response = client
+        .post(format!("{API_URL}circuit/create"))
+        .headers(header.clone())
+        .multipart(upload)
+        .send()
+        .await
+        .unwrap();
+    assert_eq!(
+        &response.status().as_u16(),
+        &201u16,
+        "Expected status code 201"
+    );
+    let response_body = response.json::<Value>().await.unwrap();
+    let circuit_id = response_body["circuit_id"].as_str().unwrap();
+    println!("Circuit ID: {:?}", &circuit_id);
+
+    // Poll circuit detail until it has a status of Ready or Failed.
+    let circuit_data = poll_circuit_status(header, circuit_id).await;
+
+    if circuit_data["status"].as_str().unwrap().contains("Failed") {
+        println!("Circuit compilation failed.");
+        std::process::exit(1);
+    }
+
+    println!("Saving circuit details locally");
+    std::fs::create_dir_all("./data").unwrap();
+    let file = File::create("./data/compile_out.json").unwrap();
+    let mut writer = BufWriter::new(file);
+    serde_json::to_writer_pretty(&mut writer, &circuit_data).unwrap();
+    writer.flush().unwrap();
+}
+
+// This function proves the circuit using the input data provided by the user and saves the proof in the /data/ directory.
```

Comment:
Perhaps this comment was added after the file was `rustfmt`ed for line length?

Can we add a `rustfmt` "check" step in the .github/workflows/ci.yaml file which verifies conformance to our accepted Rust formatting rules, whatever they are? This could be a followup task.

Reply:
I'll make a ticket to explore the feasibility of implementing such a check.  According to [this discussion ](this discussion ), there is a field called `wrap_comments` which can be set to `true` in the rustfmt config file.  However, this is not "stable" -- meaning that it's likely only available in nightly Rust.  Which is a bit od...as the initial discussion post dates back to 2016...
Adding this to my do-do list.

Reply:
committed your suggested change:  [aabc9d2](https://github.com/Sindri-Labs/sindri-resources/pull/98/commits/aabc9d200b8592ab1f62d5ec3290d9d3718723ba)